clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -373,7 +373,12 @@ public void close() throws IOException {
             }
             error += " Kafka Client will go to AUTHENTICATION_FAILED state.";
             //Unwrap the SaslException inside `PrivilegedActionException`
-            throw new SaslAuthenticationException(error, e.getCause());
+            Throwable cause = e.getCause();
+            // Treat transient Kerberos errors as non-fatal.
+            if (kerberosError != null && kerberosError.retriable)
+                throw new SaslException(error, cause);
```

Comment:
Does this make a difference? When the caller receives a SaslException, it won't mark the connection as authentication_failed, but the exception will still be propagated to the client and is not retriable?

Reply:
@junrao Thanks for the review. `SaslException` is an I/O exception and is retried similar to a network exception. `SaslAuthenticationException` is handled as a non-retriable exception that is propagated immediately to the caller.

Reply:
@rajinisivaram : Thanks for the reply. Just to clarify, at which layer is network exception (IOException) retried, selector, networkClient, or producer/consumer?

Reply:
1. `Selector` notifies disconnections including the state in which the disconnection occurred. For an I/O exception during authentication, disconnect state is `AUTHENTICATE` and for `AuthenticationException`, disconnect state is `AUTHENTICATION_FAILED`.
2. `NetworkClient` processes disconnect notifications from `Selector`. If disconnect state is `AUTHENTICATE`, it prints out a warning  but treats it like I/O exceptions in any other state. If disconnect state is `AUTHENTICATION_FAILED`, it updates the node state including the actual  `AuthenticationException`.
3. `Producer/Consumer/AdminClient` take special action based on node state when in a loop waiting for metadata, coordinator etc. Regular I/O exceptions will just result in the loop continuing to retry until timeout (e.g. send another metadata request to the least loaded broker). But if the node state is `AUTHENTICATION_FAILED`, the loop is terminated and the node's `AuthenticationException` is propagated to the application immediately.

Reply:
@junrao Thanks for the reviews. I have added comments in the test class (along with another test for the `AUTHENTICATION_FAILED` case).

Reply:
@rajinisivaram : Thanks for the explanation. Mostly makes sense to me. I can see that the consumer/admin client calls NetworkClient.authenticationException() explicitly to propagate authenticationException. On the producer side, I can see that AUTHENTICATION_FAILED is propagated to the producer on metadata request through DefaultMetadataUpdater.handleAuthenticationFailure(). However, I don't see how it is  propagated to either ApiVersionRequest or ProduceRequest. Do you know the code path that handles that?

Reply:
@junrao I don't think we handle authentication failures for every ProduceRequest. The main goal of processing authentication failures was to give immediate meaningful errors in the case of security misconfiguration (so the application sees an authentication failure immediately rather than a timeout exception after a long delay). For this case, handling metadata errors seemed sufficient (even for a `send` because that waits for metadata initially). There are other scenarios like produce requests where we could do more, but we haven't priortized that work.